rendezvous/README.md

```diff
@@ -0,0 +1,173 @@
+real-time distribution, while still providing a simple interface to
+clients.
+
+
+## The Protocol
+
+The rendezvous protocol provides facilities for real-time peer
+discovery within application specific namespaces. Peers connect to the
+rendezvous point and register their presence in one or more
+namespaces.
+
+Peers registered with the rendezvous point can be discovered by other
+nodes by querying the rendezvous point. The query specifies the
+namespace for limiting application scope and optionally a maximum
+number of peers to return. The namespace can be omitted in the query,
+which asks for all peers registered to the rendezvous point.
+
+The query can also specify a timestamp, obtained from the response to
+a previous query, such that only peers registered since the timestamp
+will be returned. This allows peers to progressively refresh their
+network view without overhead, which greatly simplifies real time
+discovery.
+
+### Registration Lifetime
+
+Registration lifetime is controlled by an optional TTL parameter in
+the `REGISTER` message.  If a TTL is specified, then the registration
+persists until the TTL expires.  If no TTL was specified, then the
+registration persists while the issuing peer is still connected to the
+rendezvous point.
+
+Peers can refresh their registrations at any time with a new
+`REGISTER` message; the TTL of the new message supersedes previous
+registrations. Peers can also cancel existing registrations at any
+time with an explicit `UNREGISTER` message.
+
+### Interaction
+
+Clients `A` and `B` connect to the rendezvous point `R` and register for namespace
+`my-app` with a `REGISTER` message:
+
+```
+A -> R: REGISTER{my-app, {QmA, AddrA}}
+B -> R: REGISTER{my-app, {QmB, AddrB}}
+```
+
+Client `C` connects and registers for namespace `another-app`:
+```
+C -> R: REGISTER{another-app, {QmC, AddrC}}
+```
+
+Another client `D` can discover peers in `my-app` by sending a `DISCOVER` message; the
+rendezvous point responds with the list of current peer reigstrations.
+```
+D -> R: DISCOVER{ns: my-app}
+R -> D: {[REGISTER{my-app, {QmA, Addr}}
+          REGISTER{my-app, {QmB, Addr}}],
+         t1}
+```
+
+If `D` wants to discover all peers registered with `R`, then it can omit the namespace
+in the query:
+```
+D -> R: DISCOVER{}
+R -> D: {[REGISTER{my-app, {QmA, Addr}}
+          REGISTER{my-app, {QmB, Addr}}
+          REGISTER{another-app, {QmC, AddrC}}],
+         t2}
+```
+
+If `D` wants to progressively poll for real time discovery, it can use
+the timestamp obtained from a previous response in order to only ask
+for new registrations.
+
+So here we consider a new client `E` registering after `t1`, and a subsequent
+query that discovers just that peer by including the timestamp:
+
+```
+E -> R: REGISTER{my-app, {QmE, AddrE}}
+D -> R: DISCOVER{ns: my-app, since: t1}
+R -> D: {[REGISTER{my-app, {QmE, AddrE}}],
+         t3}
+```
+
+### Protobuf
+
+```protobuf
+message Message {
```

Comment:
So you want a `RegistrationResponse`? 
Sure, we can do that.

Reply:
Yes. Also `DiscoverResponse` could include an optional error, too

Reply:
Ok, although it's utility would be rather limited. 
What is considered an error in this case? A non-existent namespace is equivalent to empty, and namespace name that is too large will always be empty.

Reply:
> Ok, although it's utility would be rather limited.

Let's not do this then.

> So you want a RegisterResponse?

The errors that it would throw would be:
 - `E_NAMESPACE_TOO_LONG`
 - `E_OWNERSHIP_NOT_TRUSTED` (or similar)

What do you think of that?

Reply:
> `E_NAMESPACE_TOO_LONG`

that could be generalized to `E_INVALID_NAMESPACE` to convey rejected characters for some implementation. Or it could be a different error altogether.

> `E_OWNERSHIP_NOT_TRUSTED`

Maybe a generic `E_NOT_AUTHORIZED` is more apt?

Another one we want is `E_INVALID_MULTIADDR` perhaps.

Reply:
> Maybe a generic E_NOT_AUTHORIZED is more apt?

Yes

> Another one we want is E_INVALID_MULTIADDR perhaps.

Good idea, although I don't know if we should validate those as the clients might have newer incompatible protocol tables

Reply:
Let's have `E_INVALID_PEER_INFO` to signal generic rejection of the peer info, and we can see what kind of validation makes sense.